### ISSUE TITLE
aft-mtp-mount fails under Linux 5.4: "fuse: reading device: Invalid argument"

### ISSUE THREAD

**Issue by piater** [whoozle/android-file-transfer-linux#231]
Title says it all. Debug output (manual build from current git) follows below.
Under Linux 5.3.13 it works as expected.
I am not at all sure aft-mtp-mount is at fault; Linux may well be broken, but I lack background knowledge about MTP and fuse to debug this effectively.
aft-mtp-cli works as expected, including under Linux 5.4.

```
probing device 138a:0090
error: Permission denied
probing device 1199:9079
error: Permission denied
probing device 056a:5087
error: Permission denied
probing device 04f2:b5c1
error: Permission denied
probing device 0fce:01f9
capabilities = 0x000001fd
page size = 4096
<zero-packet> 
<no-packet-size-limit> 
<bulk-scatter-gather> 
<reap-after-disconnect> 
<mmap> 
<drop-privileges> 
<unknown capability 0x180>
Device usb interface: 0:0, index: 0, enpoints: 3
read control 80 06 0300 0000
languages[4]:
00000000: 04 03 09 04                                     ....

descriptor[57]:
00000000: 12 01 10 02 00 00 00 40 ce 0f f9 01 09 04 01 02 .......@........
00000010: 03 01 09 02 27 00 01 01 04 80 fa 09 04 00 00 03 ....'...........
00000020: ff ff 00 05 07 05 81 02 00 02 00 07 05 01 02 00 ................
00000030: 02 00 07 05 82 03 1c 00 06                      .........

read control 80 06 0305 0409
interface name[8]:
00000000: 08 03 4d 00 54 00 50 00                         ..M.T.P.

FUSE library version: 2.9.9
unique: 2, opcode: INIT (26), nodeid: 0, insize: 56, pid: 0
INIT: 7.31
flags=0x03fffffb
max_readahead=0x00020000
Init: fuse proto version: 7.31, capability: 0x00000ffb, async read: 1, max readahead: 131072, max write: 131072
   INIT: 7.19
   flags=0x00000020
   max_readahead=0x00020000
   max_write=0x00100000
   max_background=0
   congestion_threshold=0
   unique: 2, success, outsize: 40
fuse: reading device: Invalid argument
```

**Comment by gr4yscale** [comment]
I can confirm this issue on Linux 5.4. To me, it seems like a Linux issue, or rather the interplay between fuse and linux.

> aft-mtp-cli works as expected
I have the same results.

Things that I've tried:

1) Downgrading `fuse2` `fuse3` and `fuse-common` packages. I am using Arch Linux. I have tried thes versions:

`pacman -U fuse-common-3.6.2-1-x86_64.pkg.tar.xz fuse3-3.6.2-1-x86_64.pkg.tar.xz fuse2-2.9.9-1-x86_64.pkg.tar.xz`
and
`pacman -U fuse2-2.9.8-1-x86_64.pkg.tar.xz fuse3-3.6.1-1-x86_64.pkg.tar.xz fuse-common-3.6.1-1-x86_64.pkg.tar.xz`

2) Checking to see if there is an older version of `android-file-transfer` to use - nothing changed recently

3) Downgrading the Linux kernel to 5.3 caused some other problems with my system, so I was unable to test

See below, the output from `aft-mtp-mount` with the `-d` and `-f` options:

```
❯ aft-mtp-mount -f -d phone
probing device 1532:0037
error: Permission denied
probing device 17cc:1330
error: Permission denied
probing device 0b05:1867
error: Permission denied
probing device 174c:2074
error: Permission denied
probing device 2717:ff48
capabilities = 0x000001fd
page size = 4096
<zero-packet> 
<no-packet-size-limit> 
<bulk-scatter-gather> 
<reap-after-disconnect> 
<mmap> 
<drop-privileges> 
<unknown capability 0x180>
Device usb interface: 0:0, index: 0, enpoints: 3
read control 80 06 0300 0000
languages[4]:
00000000: 04 03 09 04                                     ....

descriptor[80]:
00000000: 12 01 10 02 00 00 00 40 17 27 48 ff 09 04 01 02 .......@.'H.....
00000010: 03 01 09 02 3e 00 02 01 04 80 fa 09 04 00 00 03 ....>...........
00000020: ff ff 00 05 07 05 81 02 00 02 00 07 05 01 02 00 ................
00000030: 02 00 07 05 82 03 1c 00 06 09 04 01 00 02 ff 42 ...............B
00000040: 01 06 07 05 02 02 00 02 00 07 05 83 02 00 02 00 ................

read control 80 06 0305 0409
interface name[8]:
00000000: 08 03 4d 00 54 00 50 00                         ..M.T.P.

FUSE library version: 2.9.8
unique: 2, opcode: INIT (26), nodeid: 0, insize: 56, pid: 0
INIT: 7.31
flags=0x03fffffb
max_readahead=0x00020000
Init: fuse proto version: 7.31, capability: 0x00000ffb, async read: 1, max readahead: 131072, max write: 131072
   INIT: 7.19
   flags=0x00000020
   max_readahead=0x00020000
   max_write=0x00100000
   max_background=0
   congestion_threshold=0
   unique: 2, success, outsize: 40
fuse: reading device: Invalid argument

```

Everything worked beautifully before 5.4! 

@whoozle Are there any tips about how to debug this, or FUSE-based filesystems in general, that you could share?

**Comment by whoozle** [comment]
can you try running it using `sudo aft-mtp-mount`?

**Comment by gr4yscale** [comment]
@whoozle I've edited my comment just 2 minutes before your post to add some extra information, pllease check it and sorry for that!

> can you try running it using sudo aft-mtp-mount?

@whoozle I have tried this. I believe it gave the `«Device or resource busy»` or `«Device is already used by another process»` errors, but I'm away from my computer. In any case, it did not fix things. You can run the GUI and cli tools as a user.

When I checked according to [the steps in FAQ](https://github.com/whoozle/android-file-transfer-linux/blob/master/FAQ.md#im-getting-ioctl_fd-usbdevfs_claiminterface-interfacenumber-device-or-resource-busy-or-device-is-already-used-by-another-process-exceptionmessage-box-right-after-you-started-the-application-this-clearly-indicates-that-some-other-process-is-accessing-mtp-device-right-now), there was no output for PIDs of proccesses using the device.

I checked that `/dev/bus/usb/xxx/xxx` had correct ownership (username / `adbusers`) in my case - yes, I am in `adbusers` group).

I tried on 2 different machines, both recently doing full system updates, so kernel + fuse together. These are the versions:

```
[2019-12-09T12:41:02+0100] [ALPM] upgraded fuse-common (3.7.0-1 -> 3.8.0-1)
[2019-12-09T12:41:02+0100] [ALPM] upgraded fuse3 (3.7.0-1 -> 3.8.0-1)
[2019-12-18T00:13:16+0100] [ALPM] upgraded fuse-common (3.8.0-1 -> 3.9.0-1)
[2019-12-18T00:13:16+0100] [ALPM] upgraded fuse3 (3.8.0-1 -> 3.9.0-1)
 ```

**Comment by MCOfficer** [comment]
Greetings,

i stumbled over this while researching a similar issue with securefs, which also uses fuse. apparently there was a change to fuse with linux 5.4 that triggered this behavior. perhaps this helps: https://bbs.archlinux.org/viewtopic.php?pid=1876403#p1876403

**Comment by piater** [comment]
> here's how securefs fixed it: [netheril96/securefs@e5b9e5d](https://github.com/netheril96/securefs/commit/e5b9e5d41a6dcdee60ab4a18b46b9a0c6475ba21)

Thanks,  @MCOfficer! In I `#ifdef`d out the entire body of the [`Init()` method in `fuse/fuse_ll.cpp`](https://github.com/whoozle/android-file-transfer-linux/blob/2ee1f663b34c731279e4b25a58c747ee3b6f79da/fuse/fuse_ll.cpp#L434), and now aft-mtp-mount and file operations on the mounted file system work again!

This is not to imply that this is the correct fix; I did not dig for the underlying issue. But it is fit at least as a transient workaround.

**Comment by whoozle** [comment]
Thanks for fix candidate, this is really appreciated! 
Although, it's unclear whether it's fuse bug or not. 

We tested performance and it looks like big writes are improving linear read/write operations a lot, so I have probably reintroduce it later when get to the bottom of this. :)
I disabled it for now in a24ef2e5ad12574ee6370011184799e4bbcf59af

Can anyone send this on LKML, they normally don't break userspace, so this commit could possibly reverted again.

**Comment by mb720** [comment]
I can reproduce this bug on Arch Linux 5.5.3. `aft-mtp-mount -V` gives: FUSE library version: 2.9.9

Using the GUI with `android-file-transfer` works fine.

Here are the error logs for `aft-mtp-mount -d ~/android-mount`:

    probing device 056a:00e6
    error: Permission denied
    probing device 0a5c:217f
    error: Permission denied
    probing device 25a7:fa23
    error: Permission denied
    probing device 04f2:b217
    error: Permission denied
    probing device 8087:0024
    error: Permission denied
    probing device 22b8:2e82
    capabilities = 0x000001ff
    page size = 4096
    <zero-packet> 
    <bulk-continuation> 
    <no-packet-size-limit> 
    <bulk-scatter-gather> 
    <reap-after-disconnect> 
    <mmap> 
    <drop-privileges> 
    <unknown capability 0x180>
    Device usb interface: 0:0, index: 0, enpoints: 3
    read control 80 06 0300 0000
    languages[4]:
    00000000: 04 03 09 04                                     ....

    descriptor[57]:
    00000000: 12 01 00 02 00 00 00 40 b8 22 82 2e 28 02 01 02 .......@."..(...
    00000010: 03 01 09 02 27 00 01 01 00 80 fa 09 04 00 00 03 ....'...........
    00000020: ff ff 00 04 07 05 81 02 00 02 00 07 05 01 02 00 ................
    00000030: 02 00 07 05 82 03 1c 00 06                      .........

    read control 80 06 0304 0409
    interface name[8]:
    00000000: 08 03 4d 00 54 00 50 00                         ..M.T.P.

    FUSE library version: 2.9.9
    unique: 2, opcode: INIT (26), nodeid: 0, insize: 56, pid: 0
    INIT: 7.31
    flags=0x03fffffb
    max_readahead=0x00020000
    Init: fuse proto version: 7.31, capability: 0x00000ffb, async read: 1, max readahead: 131072, max write: 131072
      INIT: 7.19
      flags=0x00000020
      max_readahead=0x00020000
      max_write=0x00100000
      max_background=0
      congestion_threshold=0
      unique: 2, success, outsize: 40
    fuse: reading device: Invalid argument

And using `sudo aft-mtp-mount -d ~/android-mount`:

    probing device 056a:00e6
    capabilities = 0x000001ff
    page size = 4096
    <zero-packet> 
    <bulk-continuation> 
    <no-packet-size-limit> 
    <bulk-scatter-gather> 
    <reap-after-disconnect> 
    <mmap> 
    <drop-privileges> 
    <unknown capability 0x180>
    trying to find processes which are holding fd 3
    mapped 3 to /dev/bus/usb/002/005
    got process dir 1
    Readlink /proc/1/fd/.: Invalid argument
    Readlink /proc/1/fd/..: Invalid argument
    [a lot of lines with the same error for different processes]
    Device::Find failed:Device is already used by another process
    probing device 0a5c:217f
    capabilities = 0x000001ff
    page size = 4096
    <zero-packet> 
    <bulk-continuation> 
    <no-packet-size-limit> 
    <bulk-scatter-gather> 
    <reap-after-disconnect> 
    <mmap> 
    <drop-privileges> 
    <unknown capability 0x180>
    trying to find processes which are holding fd 3
    mapped 3 to /dev/bus/usb/001/005
    got process dir 1
    Readlink /proc/1/fd/.: Invalid argument
    Readlink /proc/1/fd/..: Invalid argument
    [a lot of lines with the same error for different processes]
    Device::Find failed:Device is already used by another process
    probing device 25a7:fa23
    capabilities = 0x000001ff
    page size = 4096
    <zero-packet> 
    <bulk-continuation> 
    <no-packet-size-limit> 
    <bulk-scatter-gather> 
    <reap-after-disconnect> 
    <mmap> 
    <drop-privileges> 
    <unknown capability 0x180>
    trying to find processes which are holding fd 3
    mapped 3 to /dev/bus/usb/001/003
    got process dir 1
    Readlink /proc/1/fd/.: Invalid argument
    Readlink /proc/1/fd/..: Invalid argument
    [a lot of lines with the same error for different processes]
    Device::Find failed:Device is already used by another process
    probing device 04f2:b217
    capabilities = 0x000001ff
    page size = 4096
    <zero-packet> 
    <bulk-continuation> 
    <no-packet-size-limit> 
    <bulk-scatter-gather> 
    <reap-after-disconnect> 
    <mmap> 
    <drop-privileges> 
    <unknown capability 0x180>
    trying to find processes which are holding fd 3
    mapped 3 to /dev/bus/usb/001/006
    got process dir 1
    Readlink /proc/1/fd/.: Invalid argument
    Readlink /proc/1/fd/..: Invalid argument
    Device::Find failed:Device is already used by another process
    probing device 8087:0024
    capabilities = 0x000001ff
    page size = 4096
    <zero-packet> 
    <bulk-continuation> 
    <no-packet-size-limit> 
    <bulk-scatter-gather> 
    <reap-after-disconnect> 
    <mmap> 
    <drop-privileges> 
    <unknown capability 0x180>
    trying to find processes which are holding fd 3
    mapped 3 to /dev/bus/usb/002/002
    got process dir 1
    Readlink /proc/1/fd/.: Invalid argument
    Readlink /proc/1/fd/..: Invalid argument
    [a lot of lines with the same error for different processes]
    Device::Find failed:Device is already used by another process
    probing device 22b8:2e82
    capabilities = 0x000001ff
    page size = 4096
    <zero-packet> 
    <bulk-continuation> 
    <no-packet-size-limit> 
    <bulk-scatter-gather> 
    <reap-after-disconnect> 
    <mmap> 
    <drop-privileges> 
    <unknown capability 0x180>
    Device usb interface: 0:0, index: 0, enpoints: 3
    read control 80 06 0300 0000
    languages[4]:
    00000000: 04 03 09 04                                     ....

    descriptor[57]:
    00000000: 12 01 00 02 00 00 00 40 b8 22 82 2e 28 02 01 02 .......@."..(...
    00000010: 03 01 09 02 27 00 01 01 00 80 fa 09 04 00 00 03 ....'...........
    00000020: ff ff 00 04 07 05 81 02 00 02 00 07 05 01 02 00 ................
    00000030: 02 00 07 05 82 03 1c 00 06                      .........

    read control 80 06 0304 0409
    interface name[8]:
    00000000: 08 03 4d 00 54 00 50 00                         ..M.T.P.

    FUSE library version: 2.9.9
    unique: 2, opcode: INIT (26), nodeid: 0, insize: 56, pid: 0
    INIT: 7.31
    flags=0x03fffffb
    max_readahead=0x00020000
    Init: fuse proto version: 7.31, capability: 0x00000ffb, async read: 1, max readahead: 131072, max write: 131072
      INIT: 7.19
      flags=0x00000020
      max_readahead=0x00020000
      max_write=0x00100000
      max_background=0
      congestion_threshold=0
      unique: 2, success, outsize: 40
    fuse: reading device: Invalid argument

**Comment by xaverdh** [comment]
Since this is apparently still broken, could you do a release containing the temporary fix?

**Comment by whoozle** [comment]
definitely will do. 4.0 will be released soon along with Zune support and loads of improvements and fixes

**Comment by anatol** [comment]
What is the state of this issue? Is it considered fixed in 4.0?

**Comment by whoozle** [comment]
I guess should be fixed in a24ef2e5ad12574ee6370011184799e4bbcf59af

**Comment by mb720** [comment]
Thanks a lot for fixing this 👍

I can confirm that `aft-mtp-mount` works now on Arch Linux 5.9.14.


`aft-mtp-mount -V` yields

    FUSE library version: 2.9.9

but that's the same FUSE version that `aft-mtp-mount` used when it didn't work earlier this year.
